Implement missing approvals label

kaspar030 opened the pull request:
Adds setting a label corresponding to the approvals status.
Supposed to be used as bors "block_label".

kaspar030 committed:
Implement missing approvals label

miri64 commented:
Can you add a note, on how to use that in the README, please?

miri64 commented:
At least some proper capitalization would be nice in the commit message. Please squash and mark ready for review.

kaspar030 commented:
- added README entry
- squashed & capitalized commit message

I still need to test what happens when a non-existant label is provided, and handle that properly. doing that now.

kaspar030 commented:
Ok, I added error handling for the non-existing label case. I think this is good to go!

error case output: https://github.com/RIOT-OS/RIOT/actions/runs/3882166703/jobs/6622157308

miri64 commented:
Sooo why is travis not running :-/?

miri64 reviewed (commented):
Ok, travis is out of order (probably need to port to GH actions here as well long term :-/), but I tested locally:

#### `flake8 check_labels.py`
```
check_labels.py:184:80: E501 line too long (126 > 79 characters)
check_labels.py:190:80: E501 line too long (122 > 79 characters)
check_labels.py:211:80: E501 line too long (87 > 79 characters)
check_labels.py:276:80: E501 line too long (82 > 79 characters)
check_labels.py:278:80: E501 line too long (104 > 79 characters)
check_labels.py:281:80: E501 line too long (102 > 79 characters)
check_labels.py:386:80: E501 line too long (101 > 79 characters)
check_labels.py:395:80: E501 line too long (80 > 79 characters)
check_labels.py:431:80: E501 line too long (83 > 79 characters)
check_labels.py:432:80: E501 line too long (80 > 79 characters)
check_labels.py:439:80: E501 line too long (82 > 79 characters)
check_labels.py:446:80: E501 line too long (81 > 79 characters)
check_labels.py:486:80: E501 line too long (104 > 79 characters)
```

#### `pylint check_labels.py`
```
************* Module check_labels
check_labels.py:496:0: R0022: Useless option value for 'disable', 'no-self-use' was moved to an optional extension, see https://pylint.pycqa.org/en/latest/whatsnew/2/2.14/summary.html#removed-checkers. (useless-option-value)
check_labels.py:504:0: R0022: Useless option value for 'disable', 'no-self-use' was moved to an optional extension, see https://pylint.pycqa.org/en/latest/whatsnew/2/2.14/summary.html#removed-checkers. (useless-option-value)
check_labels.py:184:0: C0301: Line too long (126/100) (line-too-long)
check_labels.py:190:0: C0301: Line too long (122/100) (line-too-long)
check_labels.py:278:0: C0301: Line too long (104/100) (line-too-long)
check_labels.py:281:0: C0301: Line too long (102/100) (line-too-long)
check_labels.py:386:0: C0301: Line too long (101/100) (line-too-long)
check_labels.py:486:0: C0301: Line too long (104/100) (line-too-long)
check_labels.py:94:25: C0209: Formatting a regular string which could be a f-string (consider-using-f-string)
check_labels.py:137:29: C0209: Formatting a regular string which could be a f-string (consider-using-f-string)
check_labels.py:140:25: C0209: Formatting a regular string which could be a f-string (consider-using-f-string)
check_labels.py:145:25: C0209: Formatting a regular string which could be a f-string (consider-using-f-string)
check_labels.py:160:21: C0209: Formatting a regular string which could be a f-string (consider-using-f-string)
check_labels.py:208:25: C0209: Formatting a regular string which could be a f-string (consider-using-f-string)
check_labels.py:220:22: C0209: Formatting a regular string which could be a f-string (consider-using-f-string)
check_labels.py:224:14: C0209: Formatting a regular string which could be a f-string (consider-using-f-string)
check_labels.py:233:22: C0209: Formatting a regular string which could be a f-string (consider-using-f-string)

-----------------------------------
Your code has been rated at 9.24/10

```

#### `pytest`
<details>

```
============================================================================================================ test session starts =============================================================================================================
platform linux -- Python 3.10.9, pytest-7.2.0, pluggy-1.0.0 -- /home/mlenders/Repositories/RIOT-OS/check_labels/env/bin/python
cachedir: .pytest_cache
rootdir: /home/mlenders/Repositories/RIOT-OS/check_labels, configfile: setup.cfg, testpaths: check_labels.py
plugins: cov-4.0.0
collected 93 items                                                                                                                                                                                                                           

check_labels.py::check_labels.cs_2tuple_list PASSED                                                                                                                                                                                    [  1%]
check_labels.py::check_labels.cs_string_list PASSED                                                                                                                                                                                    [  2%]
check_labels.py::check_labels.get_pull_no PASSED                                                                                                                                                                                       [  3%]
check_labels.py::check_labels.parse_condition PASSED                                                                                                                                                                                   [  4%]
check_labels.py::test_cs_string_list[-exp0] PASSED                                                                                                                                                                                     [  5%]
check_labels.py::test_cs_string_list_value_error["] PASSED                                                                                                                                                                             [  6%]
check_labels.py::test_cs_string_list_value_error['] PASSED                                                                                                                                                                             [  7%]
check_labels.py::test_cs_string_list_value_error[test,' foo bar] PASSED                                                                                                                                                                [  8%]
check_labels.py::test_cs_string_list_value_error[test," foo bar] PASSED                                                                                                                                                                [  9%]
check_labels.py::test_cs_string_list_value_error["test] PASSED                                                                                                                                                                         [ 10%]
check_labels.py::test_cs_string_list_value_error['test] PASSED                                                                                                                                                                         [ 11%]
check_labels.py::test_cs_2tuple_list_value_error[)] PASSED                                                                                                                                                                             [ 12%]
check_labels.py::test_cs_2tuple_list_value_error["] PASSED                                                                                                                                                                             [ 13%]
check_labels.py::test_cs_2tuple_list_value_error['] PASSED                                                                                                                                                                             [ 15%]
check_labels.py::test_cs_2tuple_list_value_error[(test] PASSED                                                                                                                                                                         [ 16%]
check_labels.py::test_cs_2tuple_list_value_error[(test, foobar, "foo bar")] PASSED                                                                                                                                                     [ 17%]
check_labels.py::test_get_pull_no_invalid PASSED                                                                                                                                                                                       [ 18%]
check_labels.py::test_parse_condition_invalid[foobar] PASSED                                                                                                                                                                           [ 19%]
check_labels.py::test_parse_condition_invalid[foobar>3>test] PASSED                                                                                                                                                                    [ 20%]
check_labels.py::test_check_review_approvals[value0-reviews0-True] PASSED                                                                                                                                                              [ 21%]
check_labels.py::test_check_review_approvals[value1-reviews1-False] PASSED                                                                                                                                                             [ 22%]
check_labels.py::test_check_review_approvals[value2-reviews2-True] PASSED                                                                                                                                                              [ 23%]
check_labels.py::test_check_review_approvals[value3-reviews3-False] PASSED                                                                                                                                                             [ 24%]
check_labels.py::test_check_review_approvals[value4-reviews4-False] PASSED                                                                                                                                                             [ 25%]
check_labels.py::test_check_condition PASSED                                                                                                                                                                                           [ 26%]
check_labels.py::test_check_condition_invalid PASSED                                                                                                                                                                                   [ 27%]
check_labels.py::test_check_labels[set_labels0-unset_labels0-cond_labels0-False-pull_labels0--0] PASSED                                                                                                                                [ 29%]
check_labels.py::test_check_labels[set_labels1-unset_labels1-cond_labels1-False-pull_labels1--0] PASSED                                                                                                                                [ 30%]
check_labels.py::test_check_labels[set_labels2-unset_labels2-cond_labels2-False-pull_labels2--0] PASSED                                                                                                                                [ 31%]
check_labels.py::test_check_labels[set_labels3-unset_labels3-cond_labels3-True-pull_labels3--0] PASSED                                                                                                                                 [ 32%]
check_labels.py::test_check_labels[set_labels4-unset_labels4-cond_labels4-True-pull_labels4--0] PASSED                                                                                                                                 [ 33%]
check_labels.py::test_check_labels[set_labels5-unset_labels5-cond_labels5-True-pull_labels5--0] PASSED                                                                                                                                 [ 34%]
check_labels.py::test_check_labels[set_labels6-unset_labels6-cond_labels6-False-pull_labels6--0] PASSED                                                                                                                                [ 35%]
check_labels.py::test_check_labels[set_labels7-unset_labels7-cond_labels7-False-pull_labels7--0] PASSED                                                                                                                                [ 36%]
check_labels.py::test_check_labels[set_labels8-unset_labels8-cond_labels8-False-pull_labels8--1] PASSED                                                                                                                                [ 37%]
check_labels.py::test_check_labels[set_labels9-unset_labels9-cond_labels9-True-pull_labels9--0] PASSED                                                                                                                                 [ 38%]
check_labels.py::test_check_labels[set_labels10-unset_labels10-cond_labels10-True-pull_labels10--0] PASSED                                                                                                                             [ 39%]
check_labels.py::test_check_labels[set_labels11-unset_labels11-cond_labels11-True-pull_labels11--0] PASSED                                                                                                                             [ 40%]
check_labels.py::test_check_labels[set_labels12-unset_labels12-cond_labels12-False-pull_labels12--0] PASSED                                                                                                                            [ 41%]
check_labels.py::test_check_labels[set_labels13-unset_labels13-cond_labels13-False-pull_labels13--0] PASSED                                                                                                                            [ 43%]
check_labels.py::test_check_labels[set_labels14-unset_labels14-cond_labels14-False-pull_labels14--1] PASSED                                                                                                                            [ 44%]
check_labels.py::test_check_labels[set_labels15-unset_labels15-cond_labels15-True-pull_labels15--0] PASSED                                                                                                                             [ 45%]
check_labels.py::test_check_labels[set_labels16-unset_labels16-cond_labels16-True-pull_labels16--0] PASSED                                                                                                                             [ 46%]
check_labels.py::test_check_labels[set_labels17-unset_labels17-cond_labels17-True-pull_labels17--1] PASSED                                                                                                                             [ 47%]
check_labels.py::test_check_labels[set_labels18-unset_labels18-cond_labels18-False-pull_labels18--0] PASSED                                                                                                                            [ 48%]
check_labels.py::test_check_labels[set_labels19-unset_labels19-cond_labels19-False-pull_labels19--0] PASSED                                                                                                                            [ 49%]
check_labels.py::test_check_labels[set_labels20-unset_labels20-cond_labels20-False-pull_labels20--1] PASSED                                                                                                                            [ 50%]
check_labels.py::test_check_labels[set_labels21-unset_labels21-cond_labels21-False-pull_labels21--1] PASSED                                                                                                                            [ 51%]
check_labels.py::test_check_labels[set_labels22-unset_labels22-cond_labels22-True-pull_labels22--0] PASSED                                                                                                                             [ 52%]
check_labels.py::test_check_labels[set_labels23-unset_labels23-cond_labels23-True-pull_labels23--0] PASSED                                                                                                                             [ 53%]
check_labels.py::test_check_labels[set_labels24-unset_labels24-cond_labels24-True-pull_labels24--0] PASSED                                                                                                                             [ 54%]
check_labels.py::test_check_labels[set_labels25-unset_labels25-cond_labels25-True-pull_labels25--1] PASSED                                                                                                                             [ 55%]
check_labels.py::test_check_labels[set_labels26-unset_labels26-cond_labels26-False-pull_labels26--1] PASSED                                                                                                                            [ 56%]
check_labels.py::test_check_labels[set_labels27-unset_labels27-cond_labels27-False-pull_labels27--1] PASSED                                                                                                                            [ 58%]
check_labels.py::test_check_labels[set_labels28-unset_labels28-cond_labels28-False-pull_labels28--1] PASSED                                                                                                                            [ 59%]
check_labels.py::test_check_labels[set_labels29-unset_labels29-cond_labels29-False-pull_labels29--0] PASSED                                                                                                                            [ 60%]
check_labels.py::test_check_labels[set_labels30-unset_labels30-cond_labels30-True-pull_labels30--1] PASSED                                                                                                                             [ 61%]
check_labels.py::test_check_labels[set_labels31-unset_labels31-cond_labels31-True-pull_labels31--1] PASSED                                                                                                                             [ 62%]
check_labels.py::test_check_labels[set_labels32-unset_labels32-cond_labels32-True-pull_labels32--1] PASSED                                                                                                                             [ 63%]
check_labels.py::test_check_labels[set_labels33-unset_labels33-cond_labels33-True-pull_labels33--0] PASSED                                                                                                                             [ 64%]
check_labels.py::test_check_labels[set_labels34-unset_labels34-cond_labels34-False-pull_labels34--1] PASSED                                                                                                                            [ 65%]
check_labels.py::test_check_labels[set_labels35-unset_labels35-cond_labels35-False-pull_labels35--1] PASSED                                                                                                                            [ 66%]
check_labels.py::test_check_labels[set_labels36-unset_labels36-cond_labels36-False-pull_labels36--1] PASSED                                                                                                                            [ 67%]
check_labels.py::test_check_labels[set_labels37-unset_labels37-cond_labels37-False-pull_labels37--0] PASSED                                                                                                                            [ 68%]
check_labels.py::test_check_labels[set_labels38-unset_labels38-cond_labels38-False-pull_labels38--1] PASSED                                                                                                                            [ 69%]
check_labels.py::test_check_labels[set_labels39-unset_labels39-cond_labels39-False-pull_labels39--1] PASSED                                                                                                                            [ 70%]
check_labels.py::test_check_labels[set_labels40-unset_labels40-cond_labels40-True-pull_labels40--1] PASSED                                                                                                                             [ 72%]
check_labels.py::test_check_labels[set_labels41-unset_labels41-cond_labels41-True-pull_labels41--1] PASSED                                                                                                                             [ 73%]
check_labels.py::test_check_labels[set_labels42-unset_labels42-cond_labels42-True-pull_labels42--1] PASSED                                                                                                                             [ 74%]
check_labels.py::test_check_labels[set_labels43-unset_labels43-cond_labels43-True-pull_labels43--1] PASSED                                                                                                                             [ 75%]
check_labels.py::test_check_labels[set_labels44-unset_labels44-cond_labels44-True-pull_labels44--0] PASSED                                                                                                                             [ 76%]
check_labels.py::test_check_labels[set_labels45-unset_labels45-cond_labels45-False-pull_labels45--1] PASSED                                                                                                                            [ 77%]
check_labels.py::test_check_labels[set_labels46-unset_labels46-cond_labels46-False-pull_labels46--1] PASSED                                                                                                                            [ 78%]
check_labels.py::test_check_labels[set_labels47-unset_labels47-cond_labels47-False-pull_labels47--1] PASSED                                                                                                                            [ 79%]
check_labels.py::test_check_labels[set_labels48-unset_labels48-cond_labels48-False-pull_labels48--1] PASSED                                                                                                                            [ 80%]
check_labels.py::test_check_labels[set_labels49-unset_labels49-cond_labels49-True-pull_labels49--1] PASSED                                                                                                                             [ 81%]
check_labels.py::test_check_labels[set_labels50-unset_labels50-cond_labels50-True-pull_labels50--1] PASSED                                                                                                                             [ 82%]
check_labels.py::test_check_labels[set_labels51-unset_labels51-cond_labels51-True-pull_labels51--1] PASSED                                                                                                                             [ 83%]
check_labels.py::test_check_labels[set_labels52-unset_labels52-cond_labels52-True-pull_labels52--0] PASSED                                                                                                                             [ 84%]
check_labels.py::test_check_labels[set_labels53-unset_labels53-cond_labels53-True-pull_labels53--1] PASSED                                                                                                                             [ 86%]
check_labels.py::test_check_labels[set_labels54-unset_labels54-cond_labels54-False-pull_labels54--1] PASSED                                                                                                                            [ 87%]
check_labels.py::test_check_labels[set_labels55-unset_labels55-cond_labels55-False-pull_labels55--1] PASSED                                                                                                                            [ 88%]
check_labels.py::test_check_labels[set_labels56-unset_labels56-cond_labels56-False-pull_labels56--1] PASSED                                                                                                                            [ 89%]
check_labels.py::test_check_labels[set_labels57-unset_labels57-cond_labels57-False-pull_labels57--1] PASSED                                                                                                                            [ 90%]
check_labels.py::test_check_labels[set_labels58-unset_labels58-cond_labels58-False-pull_labels58--1] PASSED                                                                                                                            [ 91%]
check_labels.py::test_check_labels[set_labels59-unset_labels59-cond_labels59-False-pull_labels59--1] PASSED                                                                                                                            [ 92%]
check_labels.py::test_check_labels[set_labels60-unset_labels60-cond_labels60-True-pull_labels60--1] PASSED                                                                                                                             [ 93%]
check_labels.py::test_check_labels[set_labels61-unset_labels61-cond_labels61-True-pull_labels61--1] PASSED                                                                                                                             [ 94%]
check_labels.py::test_check_labels[set_labels62-unset_labels62-cond_labels62-True-pull_labels62--1] PASSED                                                                                                                             [ 95%]
check_labels.py::test_check_labels[set_labels63-unset_labels63-cond_labels63-True-pull_labels63--1] PASSED                                                                                                                             [ 96%]
check_labels.py::test_check_labels[set_labels64-unset_labels64-cond_labels64-True-pull_labels64--0] PASSED                                                                                                                             [ 97%]
check_labels.py::test_check_labels[set_labels65-unset_labels65-cond_labels65-True-pull_labels65--1] PASSED                                                                                                                             [ 98%]
check_labels.py::test_main FAILED                                                                                                                                                                                                      [100%]

================================================================================================================== FAILURES ==================================================================================================================
_________________________________________________________________________________________________________________ test_main __________________________________________________________________________________________________________________

monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x7f330bbeb310>

    def test_main(monkeypatch):
        class MockArgs:
            set_labels = ["yes"]
            unset_labels = []
            cond_labels = []
            missing_approvals_label = ''
    
        class MockRepo:
            # pylint: disable=no-self-use
            def get_pull(self, pull_no):
                return pull_no
    
        class MockGithub:
            def __init__(self, *args, **kwargs):
                pass
    
            # pylint: disable=no-self-use, unused-argument
            def get_repo(self, name):
                return MockRepo()
    
        monkeypatch.setattr(argparse.ArgumentParser, "add_argument",
                            lambda *args, **kwargs: None)
        monkeypatch.setattr(argparse.ArgumentParser, "parse_args",
                            lambda *args, **kwargs: MockArgs())
        monkeypatch.setattr(github, "Github",
                            lambda *args, **kwargs: MockGithub())
        monkeypatch.setattr(sys.modules[__name__], "check_labels",
                            lambda *args, **kwargs: 0)
        monkeypatch.setattr(sys.modules[__name__], "get_pull_no",
                            lambda *args, **kwargs: 12345)
        monkeypatch.setenv("GITHUB_REPOSITORY", "test")
        monkeypatch.setenv("GITHUB_REF", "foobar")
>       assert main() == 0

check_labels.py:520: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

    def main():
        parser = argparse.ArgumentParser()
        parser.add_argument(
            'set_labels', default='', type=cs_string_list,
            help="Comma-seperated list of labels required to be set. "
                 "default: ''"
        )
        parser.add_argument(
            'unset_labels', default='', type=cs_string_list,
            help="Comma-seperated list of labels required not to be set. "
                 "default: ''"
        )
        parser.add_argument(
            'cond_labels', default='', type=cs_2tuple_list,
            help="Comma-seperated list of (label, condition) for labels "
                 "introducing a conditions. "
                 "default: ''. "
                 "Supported conditions: 'review.approvals>x' "
                 "where x is a positive number"
        )
        parser.add_argument(
            'missing_approvals_label', default='', type=str,
            help="Name of label reflecting the approval status,"
                 "will be set while approvals are missing"
                 "default: '' (no label is managed). "
        )
        args = parser.parse_args()
    
        repo_name = os.environ["GITHUB_REPOSITORY"]
        repo = github.Github(
            os.environ.get("INPUT_ACCESS_TOKEN", None)
        ).get_repo(repo_name)
    
        if args.missing_approvals_label:
            # turn label string into label object
            try:
                missing_approvals_label = repo.get_label(args.missing_approvals_label)
            except github.GithubException:
                print(f"Error getting label \"{args.missing_approvals_label}\" from github. Does it exist?")
                return 1
    
>       return check_labels(args.set_labels, args.unset_labels, args.cond_labels, missing_approvals_label,
                            repo.get_pull(get_pull_no(os.environ["GITHUB_REF"])))
E       UnboundLocalError: local variable 'missing_approvals_label' referenced before assignment

check_labels.py:281: UnboundLocalError

---------- coverage: platform linux, python 3.10.9-final-0 -----------
Name              Stmts   Miss Branch BrPart  Cover   Missing
-------------------------------------------------------------
check_labels.py     233     13    131      8    94%   19-20, 186, 190-191, 275-279, 286, 289->exit, 381, 498, 514->exit, 516->exit
-------------------------------------------------------------
TOTAL               233     13    131      8    94%
Coverage XML written to file coverage.xml

FAIL Required test coverage of 97% not reached. Total coverage: 93.68%
========================================================================================================== short test summary info ===========================================================================================================
FAILED check_labels.py::test_main - UnboundLocalError: local variable 'missing_approvals_label' referenced before assignment
======================================================================================================== 1 failed, 92 passed in 0.74s ========================================================================================================
```

</details>

You do not need to fix regressions due to bit rot caused by missing CI, but please fix the errors pointed out in the introduced lines.

miri64 commented:
I added a Github Action now and as a consequence I linted the code. Please rebase to resolve the merge conflicts.

miri64 commented:
(and trigger the action for this PR)

miri64 commented on check_labels.py:
```diff
@@ -174,14 +174,22 @@ def parse_condition(condition):
         if review.state == "APPROVED":
             approvals += 1
             if approvals > condition[1]:
+                print(f"PR#{pull} has {approvals}/{condition[1] + 1} approvals, removing label \"{missing_approvals_label}\"")
```
Existing prints use `'` within the string literals for quoting. What speaks against it here?
```suggestion
                print(f"PR#{pull} has {approvals}/{condition[1] + 1} approvals, removing label '{missing_approvals_label}'")
```

miri64 commented on check_labels.py:
```diff
@@ -174,14 +174,22 @@ def parse_condition(condition):
                 return True
+
+    if missing_approvals_label and condition[1] > 0:
+        print(f"PR#{pull} has only {approvals}/{condition[1] + 1} approvals, setting label \"{missing_approvals_label}\"")
```
```suggestion
        print(f"PR#{pull} has only {approvals}/{condition[1] + 1} approvals, setting label '{missing_approvals_label}'")
```

miri64 commented on check_labels.py:
```diff
@@ -248,14 +257,28 @@ def main():
+        try:
+            missing_approvals_label = repo.get_label(args.missing_approvals_label)
+        except github.GithubException:
+            print(f"Error getting label \"{args.missing_approvals_label}\" from github. Does it exist?")
```
```suggestion
            print(f"Error getting label '{args.missing_approvals_label}' from github. Does it exist?")
```

kaspar030 commented:
CI fails complaining missing test coverage. I don't know how to test getting/setting labels from github in the test env :/

miri64 commented:
> I don't know how to test getting/setting labels from github in the test env :/

Simple answer: You don't, it is mocked. I will provide a patch to increase the test coverage :-).

miri64 committed:
Increase test coverage
